sdks/java/io/azure/src/main/java/org/apache/beam/sdk/io/azure/blobstore/AzfsResourceId.java

```diff
@@ -0,0 +1,219 @@
+    }
+    String blobWithoutTrailingSlash = blob.substring(0, blob.length() - 1);
+    return blobWithoutTrailingSlash.substring(blobWithoutTrailingSlash.lastIndexOf('/') + 1);
+  }
+
+  // TODO: ensure that this function lines up with what the filesystem match method expects
+  @Override
+  public String toString() {
+    if (blob != null) {
+      return String.format("%s://%s/%s/%s", SCHEME, account, container, blob);
+    }
+    return String.format("%s://%s/%s/", SCHEME, account, container);
+  }
+
+  @Override
+  public boolean equals(Object obj) {
+    if (!(obj instanceof AzfsResourceId)) {
+      return false;
+    }
+    String otherBlob = ((AzfsResourceId) obj).blob;
+    boolean equalBlob = blob != null && otherBlob != null && blob.equals(otherBlob);
+    boolean noBlobs = blob == null && otherBlob == null;
+    return account.equals(((AzfsResourceId) obj).account)
+        && container.equals(((AzfsResourceId) obj).container)
+        && (equalBlob || noBlobs);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(account, container, blob);
+  }
+
+  @Override
+  public ResourceId resolve(String other, ResolveOptions resolveOptions) {
+    checkState(isDirectory(), "Expected this resource to be a directory, but was [%s]", toString());
+    // TODO: check if resolve options are an illegal name in any way, see:
+    // https://docs.microsoft.com/en-us/rest/api/storageservices/Naming-and-Referencing-Containers--Blobs--and-Metadata
+
+    if (resolveOptions == ResolveOptions.StandardResolveOptions.RESOLVE_DIRECTORY) {
+      if ("..".equals(other)) {
+        if ("/".equals(blob)) {
+          return this;
+        }
+        int parentStopsAt = blob.substring(0, blob.length() - 1).lastIndexOf('/');
+        return fromComponents(account, container, blob.substring(0, parentStopsAt + 1));
+      }
+
+      if ("".equals(other)) {
+        return this;
+      }
+
+      if (!other.endsWith("/")) {
+        other += "/";
+      }
+      if (AZFS_URI.matcher(other).matches()) {
+        return fromUri(other);
+      }
+      if (blob == null) {
+        return fromComponents(account, container, other);
+      }
+      return fromComponents(account, container, blob + other);
+    }
+
+    if (resolveOptions == ResolveOptions.StandardResolveOptions.RESOLVE_FILE) {
+      checkArgument(
+          !other.endsWith("/"), "Cannot resolve a file with a directory path: [%s]", other);
+      checkArgument(!"..".equals(other), "Cannot resolve parent as file: [%s]", other);
+      if (AZFS_URI.matcher(other).matches()) {
+        return fromUri(other);
+      }
+      if (blob == null) {
+        return fromComponents(account, container, other);
+      }
+      return fromComponents(account, container, blob + other);
+    }
+
+    throw new UnsupportedOperationException(
+        String.format("Unexpected StandardResolveOptions [%s]", resolveOptions));
+  }
+
+  // url format to interact with Azure
+  public String toAzureUrl() {
```

Comment:
It seems like this function is not used.

Reply:
I think I will need it later to interact with the Azure client, but I will remove it for now.